Replace loadpolicy with prefetch & prerender

igrigorik opened the pull request:
Preview: https://rawgit.com/w3c/resource-hints/prefetch-prerender/index.html
- removed `loadpolicy` attribute
- loadpolicy=next -> prerender
- loadpolicy=next inert -> prefetch

TODO's and open questions:
- [x] Define `as` attribute on prefetch
- [x] Define `as` on prerender? Or, should we keep it scoped to HTML documents?

igrigorik committed:
loadpolicy -> pre{fetch,render}

- removed loadpolicy attribute
- loadpolicy=next -> prerender
- loadpolicy=next inert -> prefetch

yoavweiss commented on index.html:
```diff
@@ -86,80 +91,47 @@
-            <ul>
-              <li>Resource fetches required for the next navigation SHOULD have lower relative priority and SHOULD NOT block or interfere with resource fetches required by the current navigation context.</li>
-
-              <li>The optimal time to initiate a resource fetch required for the next navigation is dependent on the negotiated transport protocol, users current connectivity profile, available device resources, and other context specific variables. The user agent is left to determine the optimal time at which to initiate the fetch - e.g. the user agent MAY decide to wait until all other download ...
```
Can we add the above paragraphs (or something similar) to the new "prefetch" section?

igrigorik replied:
It is there, see line 233 in updated file.

yoavweiss replied:
Ha, totally missed it

yoavweiss commented:
LGTM

igrigorik commented:
Any opinions / suggestions on prerender + `as`? 

Practically speaking, today nothing stops someone from specifying a non-HTML resource via `<link rel=prerender>`, it's just that existing UA's that support page prerendering won't do anything with the response. So, in effect, such a fetch is basically equivalent to `<link rel=prefetch>`.

In the spirit of keeping things simple, I propose we omit "as" on prerender and implicitly restrict it to HTML and page prerendering use cases. Nothing stops us from expanding this functionality in the future by adding `as` support. 

Thoughts?

yoavweiss commented:
Sounds good as long as the UA is free to preprocess other types of resources (without triggering extra requests or any other side effects)

igrigorik committed:
scope prerender to HTML resources

We can expand the scope to non-HTML resources, if needed, in the future.
As such, current definition specifies the currently implemented behavior
in Chrome and IE.

igrigorik committed:
define 'as' as optional for prefetch

- 'as' must be optional to be backwards compatible
- 'as' is defined in Preload, link there to avoid duplication

igrigorik committed:
cleanup 'pr' section

igrigorik committed:
move load and error down for consistency

igrigorik commented:
- I've constrained prerender to HTML only.
- I've made `as` optional on prefetch and linked it's processing/definition to Preload. (Note: it must me optional to be backwards compatible).

As such, I believe this should be good to merge now.. Assuming I get an LGTM :-)

/cc @yoavweiss @toddreifsteck @plehegar @marcoscaceres

yoavweiss commented on index.html:
```diff
@@ -86,80 +91,56 @@
+        <li>Delay some requests until the requested HTML resource is made visible - e.g. media downloads, plugin content, and so on.</li>
+        <li>Prevent prerendering from being initiated when there are limited resources available.</li>
+        <li>Abandon prendering due to high resource requirements - e.g. exceeded memory requirements, high CPU usage, and so on.</li>
+        <li>Abandon prerendering due to the type or properties of the fetched content:</li>
```
Should we add network "cost" to the list?

igrigorik replied:
Done, see 7617c67.

yoavweiss commented on index.html:
```diff
@@ -244,14 +217,30 @@
+    <section>
+      <h2>Load and error events</h2>
+
+      <p>The user agent MUST NOT <a href="https://html.spec.whatwg.org/#delay-the-load-event">delay the `load` event</a> of the document due to outstanding <code><a>preconnect</a></code> or <code><a>speculative fetch</a></code> initiated requests.</p>
```
What happens with outstanding `preload` fetches? We mustn't block on them either IMO (unless the parser encountered the elements that require them before onload)

yoavweiss replied:
Maybe it'd be easiest to add them to `speculative fetch`

igrigorik replied:
@yoavweiss preload behavior is covered in a separate spec, we'll define their behavior there... Unless, I'm missing something?

yoavweiss replied:
You're not missing anything, I was commenting before my morning tea :)

toddreifsteck commented:
Flagging for review

toddreifsteck commented:
Ok.. after a quick review, this looks good to merge. Will review in detail with sub-team that has previously implemented IE's prefetch/prerender logic and come back with any additional notes in future.

igrigorik committed:
add network cost as prerendering consideration

yoavweiss commented:
LGTM

igrigorik commented:
Merged in d33f6039be2df265135eb3c8e258423542058d9a.
